### ISSUE TITLE
Bug: Flat config file not found when run with `CWD` outside of source directory

### ISSUE THREAD

**Issue by danfuzz** [eslint/eslint#18196]
### Environment

Node version: 21.4.0
npm version: 10.2.4
Local ESLint version: N/A
Global ESLint version: 8.57.0
Operating System: macOS

### What parser are you using?

Default (Espree)

### What did you do?

I set up a small project to demonstrate a problem where ESLint does not seem to find a flat config file: [config-problem.tgz](https://github.com/eslint/eslint/files/14576805/config-problem.tgz)

If you untar this, you'll find this file structure:

```
config-problem/
  src/
    eslint.config.mjs
    a-module/
      oh-no.js
      package.json
```

The file `oh-no.js` has a problem which should get detected by the config file. If you `cd` to `src`, then ESLint will successfully report the problem:

```
$ cd config-problem/src
$ npx eslint .
/Users/danfuzz/tmp/config-problem/src/a-module/oh-no.js
  2:15  error  Missing semicolon  semi

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.

$
```

However, if you are just `cd`ed into the `config-problem` directory (above `src`), then ESLint fails to find the config file.

### What did you expect to happen?

```
$ cd config-problem
$ npx eslint src
/Users/danfuzz/tmp/config-problem/src/a-module/oh-no.js
  2:15  error  Missing semicolon  semi

✖ 1 problem (1 error, 0 warnings)
  1 error and 0 warnings potentially fixable with the `--fix` option.

$
```

### What actually happened?

```
$ npx eslint src

Oops! Something went wrong! :(

ESLint: 8.57.0

ESLint couldn't find a configuration file. To set up a configuration file for this project, please run:

    npm init @eslint/config

ESLint looked for configuration files in /Users/danfuzz/tmp/config-problem/src/a-module and its ancestors. If it found none, it then looked in your home directory.

If you think you already have a configuration file or if you need more help, please stop by the ESLint Discord server: https://eslint.org/chat

$
```

### Link to Minimal Reproducible Example

https://github.com/eslint/eslint/files/14576805/config-problem.tgz

### Participation

- [ ] I am willing to submit a pull request for this issue.

**Comment by Tanujkanti4441** [comment]
Hi @danfuzz thanks for the issue, but your link to the minimal reproduction example is not working can you fix this.

**Comment by danfuzz** [comment]
@Tanujkanti4441 Here's a gist which includes each of the files separately, along with an easy copy-paste script to recreate the example directory with all the contents: https://gist.github.com/danfuzz/132a89ee2f97e38b560a4fa084799f7c

And because it's so short, here's the actual script:

```bash
echo '
H4sIAAaa8GUAA+2V306DMBTGd81TnHCDJg5PWVuSGZ/EeIGsKBNa0kIyZ/budmxu0exP5jaMs7+
knHDa0JavX0+qZJY/9yutngpR3vbOASLGjEEb+SJiRBdxCRDKYkrIIBpEgIRyjHvAzrKabzSmTr
RdyiiRWTOdbh1nh2XZju8s97GKf4T0q/5Gp6c/A4fobx9Wf0Zx4PTvgg36J/1SjZpCnOwgHK5/T
JA7/btgg/7CFLmsw0VPWI7NsXPY/8Ep3a4/QbbWP4oBI+Qce4Cn2OA+/rn+YlIpXcNIZElT1PDg
AbzbBqDtDWCGyxeAwIgyD4YQCK2VDtrszJu3xzvvtzfh+DG77v8qSV+TZxGOjZLHzLHP/8jW/me
UW//beyB2/u+Cub99mZTCH4L/qbx/M8/Wb1WbXea8mfP55bHL/+qlL1V4/vqPtnNV/0lb/ylnzv
9dkDUyrXMlISuUrq6u24JvD4WpYQL3QNDZ3uFwOC6SDy5iIdUAFgAA' \
| base64 --decode | tar xzf -
```

**Comment by Tanujkanti4441** [comment]
I can reproduce this, the behavior is shown by `flat-config` but not by the `legacy-config`. This needs to be fixed
@eslint/eslint-team.

**Comment by mdjermanovic** [comment]
This works as intended, ESLint searches for flat config files in the current working directory and its parent directories.

You can either `cd src` and then run eslint, or use the `--config` CLI option:

https://eslint.org/docs/latest/use/configure/configuration-files-new#configuration-file-resolution
